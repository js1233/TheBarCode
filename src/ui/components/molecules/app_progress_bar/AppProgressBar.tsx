import React from "react";
import {
  StyleProp,
  StyleSheet,
  TextStyle,
  View,
  ViewStyle
} from "react-native";
import * as Progress from "react-native-progress";
import { AppLabel } from "ui/components/atoms/app_label/AppLabel";
import { FONT_SIZE } from "config/Dimens";
import { usePreferredTheme } from "hooks";

export interface AppProgressBarProps {
  style?: StyleProp<ViewStyle>;
  borderWidth?: number;
  borderRadius?: number;
  filledColor?: string;
  unFilledColor?: string;
  shouldShowBottomText?: boolean;
  bottomTextStyle?: StyleProp<TextStyle>;
  progressPercentage: number; //0 to 100
}

export const AppProgressBar = React.memo<AppProgressBarProps>(
  ({
    style,
    borderWidth = 0,
    borderRadius = 50,
    unFilledColor,
    filledColor,
    shouldShowBottomText = true,
    bottomTextStyle,
    progressPercentage
  }) => {
    const { themedColors } = usePreferredTheme();
    const progress = progressPercentage / 100;

    if (progressPercentage < 0 || progressPercentage > 100) {
      throw new Error(
        "InValid 'progressPercentage' provided to AppProgressBar "
      );
    }

    return (
      <View style={[styles.container, style]}>
        <Progress.Bar
          progress={progress}
          width={null}
          borderWidth={borderWidth}
          borderRadius={borderRadius}
          height={20}
          unfilledColor={
            unFilledColor ? unFilledColor : themedColors.interface["500"]
          }
          useNativeDriver={true}
          animationType="timing"
          color={filledColor ? filledColor : themedColors.interface["900"]}
        />
        {shouldShowBottomText && (
          <View testID="bottom-view" style={styles.textWrapper}>
            <AppLabel
              text="Complete profile & questionnaire: "
              style={[styles.textStyle, bottomTextStyle]}
            />
            <AppLabel
              text={progressPercentage + "%"}
              style={[styles.textStyle, styles.boldText, bottomTextStyle]}
            />
          </View>
        )}
      </View>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  textWrapper: {
    flexDirection: "row",
    paddingTop: 5
  },
  boldText: {
    fontWeight: "700"
  },
  textStyle: {
    fontSize: FONT_SIZE.xs,
    fontWeight: "400"
  }
});
