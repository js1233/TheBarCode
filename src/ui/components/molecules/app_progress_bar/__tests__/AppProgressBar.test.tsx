import renderer from "react-test-renderer";
import React from "react";
import { AppProgressBar } from "../AppProgressBar";
import { render } from "@testing-library/react-native";

it("Renders snapshot as expected", () => {
  // given
  const tree = renderer
    .create(<AppProgressBar progressPercentage={25} />)
    .toJSON();

  // then
  expect(tree).toMatchSnapshot();
});

it("component throw error", () => {
  // given
  const percentage = 25;
  const { getByText } = render(
    <AppProgressBar progressPercentage={percentage} />
  );
  const bottomView = getByText(percentage.toString() + "%");

  //then
  expect(bottomView).toBeDefined();
});

describe("check bottom view visibility", () => {
  it("test bottom view is visible", () => {
    // given
    const { queryByTestId } = render(
      <AppProgressBar progressPercentage={25} />
    );
    const bottomView = queryByTestId("bottom-view");

    //then
    expect(bottomView).toBeDefined();
  });

  it("test bottom view is Invisible", () => {
    // given
    const { queryByTestId } = render(
      <AppProgressBar
        progressPercentage={25}
        shouldShowBottomText={false}
      />
    );
    const bottomView = queryByTestId("bottom-view");

    //then
    expect(bottomView).toBe(null);
  });
});

it("throw error when wrong percentage is provided", () => {
  // given
  const percentage = 101;
  try {
    render(<AppProgressBar progressPercentage={percentage} />);

    expect(percentage).toBe(null);
  } catch (e: any) {
    //then
    expect(e).toBeDefined();
  }
});
