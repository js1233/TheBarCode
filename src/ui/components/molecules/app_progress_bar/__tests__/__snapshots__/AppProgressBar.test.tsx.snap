// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders snapshot as expected 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    animated={true}
    animationConfig={
      Object {
        "bounciness": 0,
      }
    }
    animationType="timing"
    indeterminate={false}
    indeterminateAnimationDuration={1000}
    onLayout={[Function]}
    progress={0.25}
    style={
      Array [
        Object {
          "backgroundColor": "#737373",
          "borderColor": "#171717",
          "borderRadius": 50,
          "borderWidth": 0,
          "overflow": "hidden",
          "width": null,
        },
        undefined,
      ]
    }
    useNativeDriver={true}
  >
    <View
      collapsable={false}
      nativeID="animatedComponent"
      style={
        Object {
          "backgroundColor": "#171717",
          "borderRadius": 50,
          "height": 20,
          "transform": Array [
            Object {
              "translateX": -0,
            },
            Object {
              "translateX": -0,
            },
            Object {
              "scaleX": 0.250075,
            },
          ],
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "paddingTop": 5,
      }
    }
    testID="bottom-view"
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "fontFamily": "ProximaNova-Regular",
          },
          Object {
            "color": "#171717",
            "fontSize": 27,
          },
          Object {
            "lineHeight": 35,
          },
          Array [
            Object {
              "fontSize": 25,
              "fontWeight": "400",
            },
            undefined,
          ],
        ]
      }
    >
      Complete profile & questionnaire: 
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "fontFamily": "ProximaNova-Regular",
          },
          Object {
            "color": "#171717",
            "fontSize": 27,
          },
          Object {
            "lineHeight": 35,
          },
          Array [
            Object {
              "fontSize": 25,
              "fontWeight": "400",
            },
            Object {
              "fontWeight": "700",
            },
            undefined,
          ],
        ]
      }
    >
      25%
    </Text>
  </View>
</View>
`;
